### DISCUSSION TITLE
how to train dreambooth on multiple GPUs?

### DISCUSSION THREAD

**Discussion by gstswwx** [d8ahazard/sd_dreambooth_extension#232] [Q&A]
Great work! Thanks for your efforts! I have a little problem: When training dreambooth, I found only 1 GPU is working, but there are 2 GPUs on the machine. I am wondering how can I use 2 GPUs to train? Much thanks! @d8ahazard

**Comment by d8ahazard** [comment]
I *think* you need to try getting deepspeed working to utilize multiple GPU's, but I'm honestly not sure.

**Reply by sgsdxzy** [reply]
Should't accelerate be enough to handle these (small-scale) parallel training? I think use accelerate-config to select multiple gpu, launch webui using accelerate, it might work out of the box.

**Reply by gstswwx** [reply]
I think deepspeed is not necessary. Without deepspeed, I could get 2 GPUs working simultaneously using example codes of diffusers official repository. @d8ahazard 

I also tried your solution. @sgsdxzy I changed the last line of `webui.sh` to `accelerate launch "${LAUNCH_SCRIPT}" "$@"`, but I found two webuis were started on the same GPU, and they listened to different ports (not the result I want）.

**Reply by sgsdxzy** [reply]
I think that's a bug for webui. AUTOMATIC1111 never had two graphics cards to test so he said he cannot fix that. You can wait until the new features made in this repo merged upstream to ShivamShrirao's diffuser and use the standalone script.

**Reply by Lukium** [reply]
> I think deepspeed is not necessary. Without deepspeed, I could get 2 GPUs working simultaneously using example codes of diffusers official repository. @d8ahazard
> 
> I also tried your solution. @sgsdxzy I changed the last line of `webui.sh` to `accelerate launch "${LAUNCH_SCRIPT}" "$@"`, but I found two webuis were started on the same GPU, and they listened to different ports (not the result I want）.

I've been working on this, you have to modify ./webui.sh | .\webui.bat because they're not setup correctly.
add `--num_processes 1` to the accelerate launch towards the very end of either file along with `--num_cpu_threads_per_process=6` as an argument.

**Reply by noxiouscardiumdimidium** [reply]
i had to also remove xformers (tesla m40 with maxwell arch, maxwell/900 series gm200 chipset does NOT play well with xformers, uses MORE vram and runs SLOWER) added export TORCH_CUDA_ARCH_LIST=5.2
export CUDA_VISIBLE_DEVICES=0,1 to webui-user.sh, and had change cross-attention analyzer to invokeai by adding --opt-split-attention-invokeai to commandline args, and disable auto1111 settings that involve any use of cpu or unloading to ram get m40+980ti "kinda" working. now all the things get auto unloaded and prefetched by invokeai and passed between cards

**Reply by Lukium** [reply]
> i had to also remove xformers (tesla m40 with maxwell arch, maxwell/900 series gm200 chipset does NOT play well with xformers, uses MORE vram and runs SLOWER) added export TORCH_CUDA_ARCH_LIST=5.2 export CUDA_VISIBLE_DEVICES=0,1 to webui-user.sh, and had change cross-attention analyzer to invokeai by adding --opt-split-attention-invokeai to commandline args, and disable auto1111 settings that involve any use of cpu or unloading to ram get m40+980ti "kinda" working. now all the things get auto unloaded and prefetched by invokeai and passed between cards

O.O !!!! Will try this shortly and report back. Thanks for the info!

**Reply by Lukium** [reply]
> i had to also remove xformers (tesla m40 with maxwell arch, maxwell/900 series gm200 chipset does NOT play well with xformers, uses MORE vram and runs SLOWER) added export TORCH_CUDA_ARCH_LIST=5.2 export CUDA_VISIBLE_DEVICES=0,1 to webui-user.sh, and had change cross-attention analyzer to invokeai by adding --opt-split-attention-invokeai to commandline args, and disable auto1111 settings that involve any use of cpu or unloading to ram get m40+980ti "kinda" working. now all the things get auto unloaded and prefetched by invokeai and passed between cards

I'm trying this now and not getting much success, do you actually see multiple GPUs being used in nvidia-smi? Did you change num_processes to 1 in `accelerate launch`?

I'm pretty much trying to get mine setup identical to yours since you had a successful setup and then optimize from there. Would you open to chat in Discord?

Thanks!

**Reply by noxiouscardiumdimidium** [reply]
that's why i said "just to get it "kinda" working", after i had it configured like that and DISABLED all the " on cpu" "unload to ram" options in settings, those functions STILL work by dumping to the 980 - both in dreambooth and while rendering. open a monitor app (im on xubuntu linux using Psensor) and watch resource usage, and do some benchmarks. when batch processing things like image captions, cahing latents in dream, any rendering process that would bring me to verge of a vram crash, both gpus hit 100% simultaneously, 980 vram use spikes, m40 vram drops, both go back to their previos load. as far as i know, the one and only way to get 100% utilazation of both is on a linux based server/crypto-miner style setup, configuered with something promox, or docker to run droneswarm/dockerswarm virtualized local cloud-computing, by deploying multiple VM's with a gpu assigned to each, both with a full install of the software and accelerate configuered with not just multi-gpu, but multiple-machine cloud-computing.

**Reply by Lukium** [reply]
> that's why i said "just to get it "kinda" working", after i had it configured like that and DISABLED all the " on cpu" "unload to ram" options in settings, those functions STILL work by dumping to the 980 - both in dreambooth and while rendering. open a monitor app (im on xubuntu linux using Psensor) and watch resource usage, and do some benchmarks. when batch processing things like image captions, cahing latents in dream, any rendering process that would bring me to verge of a vram crash, both gpus hit 100% simultaneously, 980 vram use spikes, m40 vram drops, both go back to their previos load. as far as i know, the one and only way to get 100% utilazation of both is on a linux based server/crypto-miner style setup, configuered with something promox, or docker to run droneswarm/dockerswarm virtualized local cloud-computing, by deploying multiple VM's with a gpu assigned to each, both with a full install of the software and accelerate configuered with not just multi-gpu, but multiple-machine cloud-computing.

I guess I'm just gonna have to give that a go then XD

**Reply by noxiouscardiumdimidium** [reply]
i'm capped by limitations of cpu/ram/psu. i've taken it as far as i can with what i can reasonably offord. but once configured properly virtualized/irl cloud computing is 100% doable - its how stability.ai got started without a home office, or server/super-computer. if it's for implementing THAT exact procedure ( instead of half-a$$ homebrew hackery like i'm pulling) just hit up stability.ai directly and ask about proper implementation (the people with THAT answer tend to ignore most people asking the question, because the people asking don't mean "teach me to run a cloud-compute server" therefore the answer doesn't apply xP

**Reply by noxiouscardiumdimidium** [reply]
also: another thing i forgot to mention that my hackery helps with: when rendering samples and exporting a ckpt while training an ema_only base, the vae needed to render/compile v+ema ckpt, gets loaded to "helper" gpu instead of cpu/ram, helping with system resource instability that might otherwise cause freezing/crashing . i watch 1080p video with DE at 4k without crashes now, individual iteration speeds are roughly the same but session times are shorter, because main compute gpu runs cooler, and DE/WM is more stable because more overhead core resources available - haven't crashed and had to hard-reset while exporting a ckpt since using these settings. now i need almost no optimizations (which is where i DO gain speed) unless i'm caching latents while TE+EMA training, and even then 8-bit+f16 is enough to keep me under 24gb vram

**Comment by d8ahazard** [comment]
Maybe? I can't test, so I'm honestly not sure.

**Reply by Lukium** [reply]
I'm happy to help test on my current 4 GPU dev environment. It's currently being setup to run Dedicated instances for community members, meaning that much of the time multiple GPUs will be idle and available for running tests.

**Comment by Lukium** [comment]
Has anyone been able to get accelerate to work to use multiple GPUs for training yet?

I've been trying to setup different options in `accelerate config` unsuccessfully

**Reply by bio-punk** [reply]
I have 8 4090, when I run in 1 gpu, everything is ok, but GPU out of memory when I use 8 gpu. my scripts is

```
export NCCL_DEBUG=INFO
export NCCL_IB_DISABLE=0
export NCCL_IB_HCA=mlx5_bond_0
export NCCL_SOCKET_IFNAME=bond0
export NCCL_IB_GID_INDEX=3

#export PYTORCH_CUDA_ALLOC_CONF=max_split_size_mb:20480

module load anaconda/2020.11
source activate pytorch2.0

export MODEL_NAME="/HOME/scw6898/run/.cache/huggingface/hub/stable-diffusion-v1-5"
export INSTANCE_DIR="./dog"
export OUTPUT_DIR="path_to_saved_model"

accelerate launch \
  train_dreambooth.py \
    --pretrained_model_name_or_path=$MODEL_NAME  \
    --instance_data_dir=$INSTANCE_DIR \
    --output_dir=$OUTPUT_DIR \
    --instance_prompt="a photo of sks dog" \
    --resolution=412 \
    --train_batch_size=1 \
    --gradient_accumulation_steps=1 \
    --learning_rate=5e-6 \
    --lr_scheduler="constant" \
    --lr_warmup_steps=0 \
    --max_train_steps=4000 \
    --mixed_precision=fp16 \
    # --push_to_hub

```
